### PR TITLE
Hw20

### PR DESCRIPTION
**Pull request by nightdread**

### CODE REVIEW

File: hw20/hw20.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+---
+#My first playbook
+- name: First playbook
+  hosts: linux
+  
+  tasks:
+   - import_role:
```
Comment (mavrik90) [inline hw20/hw20.yml]: Why `import_role` here?

File: hw20/hw20.yml [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+      name: hw201
```
Comment (mavrik90) [inline hw20/hw20.yml]: New line. Fix in all files please

File: hw20/roles/hw201/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,22 @@
+---
+#For CentOS
+- name: Install programs on CentOS
+  import_tasks: centos.yml
+  when: ansible_os_family == "CentOS"
+#For Ubuntu
```
Comment (mavrik90) [inline hw20/roles/hw201/tasks/main.yml]: Use empty lines between tasks

File: hw20/roles/hw201/tasks/ubuntu.yml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+---
+- name: "Install programs on Ubuntu"
```
Comment (mavrik90) [inline hw20/roles/hw201/tasks/ubuntu.yml]: Add `when` condition for ubuntu to be more flexible

Comment (nightdread) [reply hw20/roles/hw201/tasks/ubuntu.yml]: complete

File: hw20/roles/hw201/tasks/main.yml (hunk 3 above)
Comment (nightdread) [reply hw20/roles/hw201/tasks/main.yml]: complete

File: hw20/hw20.yml (hunk 2 above)
Comment (nightdread) [reply hw20/hw20.yml]: complete

File: hw20/hw20.yml (hunk 1 above)
Comment (nightdread) [reply hw20/hw20.yml]: fixed

File: hw20/roles/hw201/tasks/centos.yml [hunk 5]
```diff
@@ -1,12 +1,13 @@
 ---
 - name: "Install programs on CentOS"
-  ansible.builtin.yum:
+  ansible.builtin.dnf:
```
Comment (andreivaloshyn) [inline hw20/roles/hw201/tasks/centos.yml]: Did you try using package

File: hw20/roles/hw201/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+- name: Install programs on Ubuntu
+  import_tasks: ubuntu.yml
+  
+#Install docker
+- name: Install docker
+  import_tasks: docker.yml 
+   
+#Add user Ubuntu
+- name: Add the user 'hw20' with a sh shell and '/opt/hw20' home directory Ubuntu
```
Comment (mavrik90) [inline hw20/roles/hw201/tasks/main.yml]: User state?

Comment (nightdread) [reply hw20/roles/hw201/tasks/main.yml]: fix

File: hw20/roles/hw202/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,41 @@
+---
+#Copy to Ubuntu
+- name: "Copy files to hosts"
+  become: true
+  become_user: root
+  ansible.builtin.copy:
+   src: /home/nightdread/tmsrep/hw20/roles/hw202/files/curl.deb
```
Comment (mavrik90) [inline hw20/roles/hw202/tasks/main.yml]: use relevant path

File: hw20/roles/hw201/tasks/template.yml [hunk 8]
```diff
@@ -0,0 +1,10 @@
+---
+
+- name: "Template a file to /tmp/HW20_FILE_INFO"
+  ansible.builtin.template:
+    src: /home/nightdread/tmsrep/hw20/roles/hw201/templates/main.j2
```
Comment (mavrik90) [inline hw20/roles/hw201/tasks/template.yml]: use relevant path

File: hw20/roles/hw202/tasks/main.yml (hunk 7 above)
Comment (nightdread) [reply hw20/roles/hw202/tasks/main.yml]: fix

File: hw20/roles/hw201/tasks/template.yml (hunk 8 above)
Comment (nightdread) [reply hw20/roles/hw201/tasks/template.yml]: fix

File: hw20/group_vars/centos/centos.yml [hunk 9]
```diff
@@ -0,0 +1,5 @@
+---
+#Centos vars
+ansible_user: vagrant
```
Comment (mavrik90) [inline hw20/group_vars/centos/centos.yml]: `ansible_user` and `ansible_ssh_private_key_file` should be in general var group as you have same values.

Comment (mavrik90) [reply hw20/group_vars/centos/centos.yml]: You can leave it only in `linux` or create `all`

File: hw20/roles/hw201/tasks/centos.yml [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+  ansible.builtin.package:
+   name: "{{ item }}"
+   state: latest
+   update_cache: yes
+  loop:
+   - git
+   - python3
+   
+  become: true
+  when: ansible_facts['os_family'] == "RedHat"
```
Comment (mavrik90) [inline hw20/roles/hw201/tasks/centos.yml]: Do you need this here?

File: hw20/roles/hw201/tasks/docker.yml [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+      ansible.builtin.apt_key:
+       url: https://download.docker.com/linux/ubuntu/gpg
+       state: present
+       
+    - name: "Add repos"
+      ansible.builtin.apt_repository:
+       repo: deb https://download.docker.com/linux/ubuntu focal stable
+       state: present
+       
+    - name: "Install Docker Engine"
```
Comment (mavrik90) [inline hw20/roles/hw201/tasks/docker.yml]: Duplicate 1

File: hw20/roles/hw201/tasks/docker.yml [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+#CentOS
+- name: "Install Docker on Ubuntu"
+  block:
+    - name: "Add Docker repo"
+      ansible.builtin.get_url:
+       url: https://download.docker.com/linux/centos/docker-ce.repo
+       dest: /etc/yum.repos.d/docer-ce.repo
+      
+    - name: "Install Docker"
```
Comment (mavrik90) [inline hw20/roles/hw201/tasks/docker.yml]: Duplicate 2

File: hw20/roles/hw201/tasks/docker.yml [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+    - name: "Install Docker Engine"
+      ansible.builtin.package:
+       name: "{{item}}"
+       update_cache: yes
+      loop:
+       - docker-ce
+       - docker-ce-cli
+       - containerd.io
+             
+    - name: "Add Docker to start up"
```
Comment (mavrik90) [inline hw20/roles/hw201/tasks/docker.yml]: Duplicate docker starup

File: hw20/hw20.yml [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+- hosts: linux
+  tags: hw201
+  roles:
+   - hw201
+   
+- hosts: linux
+  tags: hw202
+  roles:
+   - hw202
+
```
Comment (mavrik90) [inline hw20/hw20.yml]: new line

File: hw20/roles/hw201/tasks/install_app.yml [hunk 15]
```diff
@@ -0,0 +1,10 @@
+---
+- name: "Install Git & Python3 on Linux"
+  ansible.builtin.package:
+   name: "{{ item }}"
+   state: latest
+   update_cache: yes
+  loop:
```
Comment (mavrik90) [inline hw20/roles/hw201/tasks/install_app.yml]: put this loop into variable

File: hw20/roles/hw201/tasks/main.yml [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+#Install docker
+- name: Install docker
+  import_tasks: docker.yml
+   
+#Add user on hosts
+- name: Add the user 'hw20' with a sh shell and '/opt/hw20' home directory
+  ansible.builtin.user:
+   name: hw20
+   shell: /bin/sh
+   home: /opt/hw20
```
Comment (mavrik90) [inline hw20/roles/hw201/tasks/main.yml]: set user home as variable

File: hw20/roles/hw201/tasks/main.yml [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+  
+#Install docker
+- name: Install docker
+  import_tasks: docker.yml
+   
+#Add user on hosts
+- name: Add the user 'hw20' with a sh shell and '/opt/hw20' home directory
+  ansible.builtin.user:
+   name: hw20
+   shell: /bin/sh
```
Comment (mavrik90) [inline hw20/roles/hw201/tasks/main.yml]: shell as variable

File: hw20/roles/hw201/tasks/main.yml [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+  import_tasks: install_app.yml
+  
+#Install docker
+- name: Install docker
+  import_tasks: docker.yml
+   
+#Add user on hosts
+- name: Add the user 'hw20' with a sh shell and '/opt/hw20' home directory
+  ansible.builtin.user:
+   name: hw20
```
Comment (mavrik90) [inline hw20/roles/hw201/tasks/main.yml]: name as variable

File: hw20/roles/hw202/tasks/main.yml [hunk 19]
```diff
@@ -0,0 +1,44 @@
+---
+#Copy to Ubuntu
```
Comment (mavrik90) [inline hw20/roles/hw202/tasks/main.yml]: use variables to do copy in single task

File: hw20/roles/hw202/tasks/main.yml [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+#Install Curl on CentOS
+- name: "Install a .rpm package"
+  become: true
+  ansible.builtin.yum:
+    name: /tmp/curl.rpm
+    state: present
+    update_cache: yes
+    allow_downgrade: yes
+  when: ansible_facts['os_family'] == "RedHat"
+
```
Comment (mavrik90) [inline hw20/roles/hw202/tasks/main.yml]: new line

File: hw20/hw20.yml (hunk 14 above)
Comment (nightdread) [reply hw20/hw20.yml]: fix

File: hw20/roles/hw201/tasks/install_app.yml (hunk 15 above)
Comment (nightdread) [reply hw20/roles/hw201/tasks/install_app.yml]: fix

File: hw20/roles/hw201/tasks/main.yml (hunk 16 above)
Comment (nightdread) [reply hw20/roles/hw201/tasks/main.yml]: fix

File: hw20/roles/hw201/tasks/main.yml (hunk 17 above)
Comment (nightdread) [reply hw20/roles/hw201/tasks/main.yml]: fix

File: hw20/roles/hw201/tasks/main.yml (hunk 18 above)
Comment (nightdread) [reply hw20/roles/hw201/tasks/main.yml]: fix

File: hw20/roles/hw202/tasks/main.yml (hunk 19 above)
Comment (nightdread) [reply hw20/roles/hw202/tasks/main.yml]: fix

File: hw20/roles/hw202/tasks/main.yml (hunk 20 above)
Comment (nightdread) [reply hw20/roles/hw202/tasks/main.yml]: fix

File: hw20/hosts.yml [hunk 21]
```diff
@@ -0,0 +1,14 @@
...
+
+all:
+ children:
+  linux:
+   hosts:
+    192.168.1.200:
+    192.168.1.201:
+  ubuntu:
+   hosts:
+    192.168.1.200
```
Comment (mavrik90) [inline hw20/hosts.yml:7]: Use hosts Its only once. Use group name instead of host IP

Comment (mavrik90) [reply hw20/hosts.yml:7]: For example
```
all:
  children:
    testhost1:
      hosts:
        172.30.128.3
    testhost2:
      hosts:
        172.30.128.4
    centos:
      hosts:
        172.30.128.5
    mix:
      children:
        testhost1:
        centos:

File: hw20/roles/hw201/tasks/docker.yml [hunk 22]
```diff
@@ -0,0 +1,55 @@
+---
+#Ubuntu
+- name: "Install Docker on Ubuntu"
+  block:
+    - name: "Set up the repository"
+      ansible.builtin.package:
+       name: "{{ item }}"
+       update_cache: yes
+      loop:
```
Comment (mavrik90) [inline hw20/roles/hw201/tasks/docker.yml]: Use variable here. This task can be applied for both Ubuntu and Centos

File: hw20/roles/hw201/tasks/template.yml [hunk 23]
```diff
@@ -0,0 +1,9 @@
+---
+
+- name: "Template a file to /tmp/HW20_FILE_INFO"
```
Comment (mavrik90) [inline hw20/roles/hw201/tasks/template.yml:3]: use loop here

File: hw20/roles/hw202/tasks/main.yml [hunk 24]
```diff
@@ -0,0 +1,32 @@
...
+    update_cache: yes
+    force: yes
+    
+  when: ansible_facts['os_family'] == "Debian"
+  
+#Install Curl on CentOS
+- name: "Install a .rpm package"
+  become: true
+  ansible.builtin.yum:
+    name: /tmp/curl.rpm
```
Comment (mavrik90) [inline hw20/roles/hw202/tasks/main.yml]: Use loop here

File: hw20/hosts.yml (hunk 21 above)
Comment (nightdread) [reply hw20/hosts.yml:7]: fix

File: hw20/roles/hw201/tasks/docker.yml (hunk 22 above)
Comment (nightdread) [reply hw20/roles/hw201/tasks/docker.yml]: fix

File: hw20/roles/hw201/tasks/template.yml (hunk 23 above)
Comment (nightdread) [reply hw20/roles/hw201/tasks/template.yml:3]: fix

File: hw20/roles/hw202/tasks/main.yml (hunk 24 above)
Comment (nightdread) [reply hw20/roles/hw202/tasks/main.yml]: fix

File: hw20/roles/hw201/tasks/docker.yml [hunk 25]
```diff
@@ -0,0 +1,48 @@
...
+       
+    - name: "Add repos"
+      ansible.builtin.apt_repository:
+       repo: deb https://download.docker.com/linux/ubuntu focal stable
+       state: present
+  when: ansible_facts['os_family'] == "Debian"
+  become: true
+
+#CentOS
+- name: "Install Docker on CentOS"
```
Comment (mavrik90) [inline hw20/roles/hw201/tasks/docker.yml]: No need to do block for only one task

File: hw20/roles/hw201/tasks/hw21_user.yml [hunk 26]
```diff
@@ -0,0 +1,15 @@
+#Add user on hosts
+- name: Add the user 'hw21_1 (hw21_2)' with a sh shell and '/opt/hw21' home directory
+  ansible.builtin.user:
+   name: "{{ item.name }}"
+   shell: "{{ item.shell }}"
+   home: "{{ item.home }}"
+   groups: "{{ item.groups }}"
+   state: present
+  with_items: "{{ hw21_user }}"
+  no_log: false
```
Comment (mavrik90) [inline hw20/roles/hw201/tasks/hw21_user.yml]: Why `false` here?

File: hw20/roles/hw201/tasks/main.yml [hunk 27]
```diff
@@ -0,0 +1,37 @@
...
+  import_tasks: template.yml
+  tags:
+   - info
+
+#Test user
+- name: "Test user create"
+  import_tasks: hw21_user.yml
+  tags:
+   - hw21
+
```
Comment (mavrik90) [inline hw20/roles/hw201/tasks/main.yml]: lines

File: hw20/roles/hw202/tasks/main.yml [hunk 28]
```diff
@@ -0,0 +1,33 @@
+---
+#Copy to hosts
+- name: "Copy files to hosts"
+  become: true
```
Comment (mavrik90) [inline hw20/roles/hw202/tasks/main.yml:11]: Do you need `root` to copy files to tmp?

File: hw20/group_vars/ubuntu/ubuntu.yml [hunk 29]
```diff
@@ -0,0 +1,11 @@
...
+#Ubuntu vars
+group_hw20: ZONE_1
+sudo_groups: sudo,docker
+user_name_hw22: "hw21_1"
+hw21_user:
+ - { name: hw21_1, groups: "sudo, docker", shell: /bin/sh, home: /opt/hw21 }
+curl_files:
+ - { src: "curl.deb", dest: "/tmp/curl.deb", owner: "hw20", mode: "0744" }
+hw20_user:
+ - { name: hw20, shell: /bin/sh, home: /opt/hw20, groups: "sudo, docker" }
```
Comment (mavrik90) [inline hw20/group_vars/ubuntu/ubuntu.yml]: better to keep the same order for parameters

Comment (mavrik90) [reply hw20/group_vars/ubuntu/ubuntu.yml]: { name: hw21_1, groups: "sudo, docker", shell: /bin/sh, home: /opt/hw21 }
{ name: hw20, shell: /bin/sh, home: /opt/hw20, groups: "sudo, docker" }

will be easier to read

File: hw20/roles/hw201/tasks/hw21_user.yml (hunk 26 above)
Comment (nightdread) [reply hw20/roles/hw201/tasks/hw21_user.yml]: I have a mistake

File: hw20/roles/hw201/tasks/docker.yml (hunk 25 above)
Comment (nightdread) [reply hw20/roles/hw201/tasks/docker.yml]: fix

File: hw20/roles/hw201/tasks/hw21_user.yml (hunk 26 above)
Comment (nightdread) [reply hw20/roles/hw201/tasks/hw21_user.yml]: fix

File: hw20/roles/hw201/tasks/main.yml (hunk 27 above)
Comment (nightdread) [reply hw20/roles/hw201/tasks/main.yml]: fix

File: hw20/roles/hw202/tasks/main.yml (hunk 28 above)
Comment (nightdread) [reply hw20/roles/hw202/tasks/main.yml:11]: no. i fix that issue.

File: hw20/group_vars/ubuntu/ubuntu.yml (hunk 29 above)
Comment (nightdread) [reply hw20/group_vars/ubuntu/ubuntu.yml]: fix

File: hw20/roles/hw201/tasks/hw20_user.yml [hunk 30]
```diff
@@ -0,0 +1,20 @@
...
+  set_fact: sudo="{{ sudo_groups }}"
+  when: sudo_groups is defined
+
+#Add user on hosts
+- name: Add the user 'hw20' with a sh shell and '/opt/hw20' home directory
+  ansible.builtin.user:
+   name: "{{ item.name }}"
+   shell: "{{ item.shell }}"
+   home: "{{ item.home }}"
+   groups: "{{ item.groups }},{{ sudo }}"
```
Comment (mavrik90) [inline hw20/roles/hw201/tasks/hw20_user.yml:13]: The only thing. Better to include `,` inside sudo var to have this `"{{ item.groups }}{{ sudo }}"` where `sudo =",{{ sudo_groups }}"`

**Review by mavrik90 — APPROVED** [review]
Well done!

**Pull request merged**